### ISSUE TITLE
Pass id of generated object to SubFactory

### ISSUE THREAD

**Issue by vovean** [FactoryBoy/factory_boy#763]
I have the next piece of code:
```
deliverer = factory.SubFactory(
        'orders.tests.factories.DelivererFactory',
        company_made_id=factory.SelfAttribute('..company_made_id'),
        user_made_id=factory.PostGeneration(lambda user, *args, **kwargs: user.pk)
    )
```

`DelivererFactory` is based on the `Deliverer` object which has a `user_made_id' field. Also, every user has at least one deliverer related to it, and, this deliverer's `user_made_id` should be assigned to the `id` of the user it is assigned to. So, I am trying to pass the user's `id` as a parameter to SubFactory. Obviously, the user instance, created by the current factory should be saved first.
And that is where the mistake occurs:
```
orders/tests/tests_factory.py:9: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../../venvs/lk-szg/lib/python3.8/site-packages/factory/base.py:46: in __call__
    return cls.create(**kwargs)
../../../venvs/lk-szg/lib/python3.8/site-packages/factory/base.py:564: in create
    return cls._generate(enums.CREATE_STRATEGY, kwargs)
../../../venvs/lk-szg/lib/python3.8/site-packages/factory/django.py:141: in _generate
    return super(DjangoModelFactory, cls)._generate(strategy, params)
../../../venvs/lk-szg/lib/python3.8/site-packages/factory/base.py:501: in _generate
    return step.build()
../../../venvs/lk-szg/lib/python3.8/site-packages/factory/builder.py:272: in build
    step.resolve(pre)
../../../venvs/lk-szg/lib/python3.8/site-packages/factory/builder.py:221: in resolve
    self.attributes[field_name] = getattr(self.stub, field_name)
../../../venvs/lk-szg/lib/python3.8/site-packages/factory/builder.py:372: in __getattr__
    value = value.evaluate(
../../../venvs/lk-szg/lib/python3.8/site-packages/factory/declarations.py:321: in evaluate
    return self.generate(step, defaults)
../../../venvs/lk-szg/lib/python3.8/site-packages/factory/declarations.py:411: in generate
    return step.recurse(subfactory, params, force_sequence=force_sequence)
../../../venvs/lk-szg/lib/python3.8/site-packages/factory/builder.py:233: in recurse
    return builder.build(parent_step=self, force_sequence=force_sequence)
../../../venvs/lk-szg/lib/python3.8/site-packages/factory/builder.py:254: in build
    pre, post = parse_declarations(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

decls = {'company_made_id': <SelfAttribute('company_made_id', default=<class 'factory.declarations._UNSPECIFIED'>)>, 'name': 'Ivan', 'rating__work_access': False, 'user_made_id': <factory.declarations.PostGeneration object at 0x7f79594f5be0>}
base_pre = <DeclarationSet: {'company_made_id': 1, 'user_made_id': 1, 'created_date_time': <factory.declarations.LazyFunction obj... <factory.faker.Faker object at 0x7f79594fcd00>, 'rating': <factory.declarations.SubFactory object at 0x7f79594fcdc0>}>
base_post = <DeclarationSet: {'vehicles': <factory.declarations.PostGeneration object at 0x7f79594fc880>, 'drivers': <factory.declarations.PostGeneration object at 0x7f79594fc970>}>

    def parse_declarations(decls, base_pre=None, base_post=None):
        pre_declarations = base_pre.copy() if base_pre else DeclarationSet()
        post_declarations = base_post.copy() if base_post else DeclarationSet()
    
        # Inject extra declarations, splitting between known-to-be-post and undetermined
        extra_post = {}
        extra_maybenonpost = {}
        for k, v in decls.items():
            if enums.get_builder_phase(v) == enums.BuilderPhase.POST_INSTANTIATION:
                if k in pre_declarations:
                    # Conflict: PostGenerationDeclaration with the same
                    # name as a BaseDeclaration
>                   raise errors.InvalidDeclarationError(
                        "PostGenerationDeclaration %s=%r shadows declaration %r"
                        % (k, v, pre_declarations[k])
                    )
E                   factory.errors.InvalidDeclarationError: PostGenerationDeclaration user_made_id=<factory.declarations.PostGeneration object at 0x7f79594f5be0> shadows declaration DeclarationWithContext(name='user_made_id', declaration=1, context={})

../../../venvs/lk-szg/lib/python3.8/site-packages/factory/builder.py:171: InvalidDeclarationError
```

As for now, I have come up with the next solution:
```
@factory.post_generation
    def deliverer(self, create, extracted, **kwargs):
        self.deliverer = DelivererFactory(
            company_made_id=self.company_made_id,
            user_made_id=self.pk,
            **kwargs
        )
```
It works perfectly fine, however, it feels like it is not the best solution and the one I was trying before should do the job. So, is it even possible to do it the way I was trying to?

**Comment by rbarrois** [comment]
It's a bit hard to follow without the models.
If I understand correctly, you have `Deliverer.user_made_id` points to a `User`, and `User.deliverer` points to a `Deliverer`, with `user.deliverer.user_made_id == user.pk`; and `Deliverer.user_made_id` is a mandatory field on `Deliverer`.

In that case, you need to first create the `User`, then its `Deliverer` — and maybe attach it.

An option would be:

```python
class UserFactory(factory.django.DjangoModelFactory):
    class Meta:
        model = User

    # Other fields here

    # Ask to create a `Deliverer` *after* the user, and pass the current
    # object as the `user_made` field
    deliverer = factory.RelatedFactory(
        DelivererFactory,
        'user_made',
        company_made_id=factory.SelfAttribute('..company_made_id'),
    )

class DelivererFactory(factory.django.DjangoModelFactory):
    class Meta:
        model = Deliverer

    class Params:
        # A parameter that callers of the factory may provide, but won't be passed
        # to the Django model
        user_made = None

    user_made_id = factory.SelfAttribute('user_made.pk')
```

**Comment by vovean** [comment]
Here the short version of models:
```
class AppUser(models.Model):
    ...
    deliverer = models.OneToOneField(
        Deliverer, on_delete=models.SET_NULL, null=True, related_name="user", blank=True
    )
    ...

class Deliverer(models.Model):
    ...
    user_made_id = models.IntegerField(blank=True, null=True)
    ...
```

And it is always so, that, if `user.id == user.deliverer.user_made_id`

About your proposed solution. Next code:
```
class CompanyMadeModelFactory(DjangoModelFactory):
    company_made_id = 1
    user_made_id = 1
    created_date_time = factory.LazyFunction(datetime.now)


class AppUserFactory(CompanyMadeModelFactory):
    class Meta:
        model = 'orders.AppUser'
        django_get_or_create = ("username",)

    user_made_id = None #user has no user_made_id
    deliverer = factory.RelatedFactory(
        get_factory('DelivererFactory'),  # returns "orders.tests.factories.DelivererFactory
        'user_made',
        company_made_id=factory.SelfAttribute('..company_made_id'),
    )

class DelivererFactory(CompanyMadeModelFactory):
    class Meta:
        model = 'orders.Deliverer'

    class Params:
        # A parameter that callers of the factory may provide, but won't be passed
        # to the Django model
        user_made = None

    user_made_id = factory.SelfAttribute('user_made.id')
```

However, when I try to mention the deliverer in tests:
```
@pytest.mark.django_db
def test_factory_nested():
    deliverer = AppUserFactory(
        deliverer__name="Deli"
    ).deliverer
    assert deliverer.user_made_id == deliverer.user.id
```

It gives me `AttributeError: 'NoneType' object has no attribute 'user_made_id'`
However, what is interesting, the Deliverer gets created because when I run `assert Deliverer.objects.first().user_made_id == AppUser.objects.first().id` it passes. So, the issue here is that it doesn't get assigned to the user.
Also, I tried each of next:
```
user_made_id = factory.SelfAttribute('user_made.id')
    user_made_id = factory.SelfAttribute('..user_made.id')
    user_made_id = factory.SelfAttribute('..Params.user_made.id')
    user_made_id = factory.SelfAttribute('Params.user_made.id')
    # Just to know if it is even here
    user_made_id = factory.LazyAttribute(
        lambda self: [print("lol", self.user_made), self.user_made.pk][1]
    )
```
